ardupilot/equipment/power/20010.BatteryContinuous.uavcan

```diff
@@ -0,0 +1,35 @@
+#
+# Battery data to be sent continuously
+#
+
+float16 temperature_cells       # [C]  : Pack mounted thermistor (preferably installed between cells), NAN: field not provided
+float16 temperature_pcb         # [C]  : Battery PCB temperature (likely output FET(s) or current sense resistor), NAN: field not provided
+float16 temperature_other       # [C]  : Application dependent, NAN: field not provided
+float32 current                 # [A]  : Positive: defined as a discharge current. Negative: defined as a charging current, NAN: field not provided
+float16 voltage                 # [V]  : Battery voltage
+uint8 state_of_charge           # [%]  : The estimated state of charge, in percent remaining (0 - 100).
+uint8 slot_id                   #      : The physical location of the battery on the aircraft. 0: field not provided
+float32 capacity_consumed       # [Ah] : This is either the consumption since power-on or since the battery was full, depending on the value of STATUS_FLAG_CAPACITY_RELATIVE_TO_FULL, NAN: field not provided
+float32 capacity_remaining      # [Ah] : If STATUS_FLAG_CAPACITY_RELATIVE_TO_FULL is unset, this value is based on the assumption the battery was full when the system was powered, NAN: field not provided
+float32 full_charge_capacity    # [Ah] : Predicted battery capacity when fully charged (accounting for battery degradation), NAN: field not provided
```

Comment:
I'm sure we discussed this at some point .....

Perhaps this should be moved into BATTERY_INFO? It won't change that often. 

```suggestion
float32 full_charge_capacity    # [Ah] : Predicted battery capacity when fully charged (accounting for battery degradation), NAN: field not provided
```

Reply:
Looking back at these I am now confused with these three fields.. I think one of them is not needed
```
float32 capacity_consumed
float32 capacity_remaining
float32 full_charge_capacity
```
**Power monitor:**
`capacity_consumed` = Ah consumed since power on
However with a power module we have to assume we start full anyway, so this really is just 
`full_charge_capacity` - `capacity_remaining`

**Smart battery**
`capacity_consumed` = Total consumed -- again it is `full_charge_capacity` - `capacity_remaining`

So `capacity_consumed` is unneeded since it can be derived, and the derivation is the same in both cases.

**Proposed**
```
float32 full_charge_capacity
float32 capacity_remaining
```
If STATUS_FLAG_CAPACITY_RELATIVE_TO_FULL is set, then `capacity_remaining` assumes the battery was fully charged at boot. If unset, it is the true remaining capacity.

Reply:
Good point, what about
```
float32 full_charge_capacity
float32 capacity_consumed
```
**power module:** only reports capacity_consumed and sets full_charge_capacity=NAN
**smart battery:** reports capacity consumed as capacity relative to full (and the flag is set). Always reports full_charge_capacity, therefore remaining is `full` - `consumed`

Reply:
MAVLink did it like this in the battery message:

```
float32 capacity_consumed
float32 capacity_remaining
status flag values_relative_to_full
```

For a power monitor the `values_relative_to_full` is `false` and `capacity_consumed` is calculated since boot. The capacity remaining is a guess by the autopilot based on its assumptions about the original full state of the battery.

For a smart battery the assumption is that values_relative_to_full is `true` and the consumed and remaining values are relative to full. So you can always get the full batter capacity by adding them.

In addition, the SMART_BATTERY_INFO explicitly contains the capacity_full_specification and capacity_full (with degradation). 

Don't know where that gets you, but it is important to remember that if power monitors are of interest the only thing you can trust is capacity consumed. Everything else is a guess. So I think that should always be sent.

Reply:
I see no point in sending `full_charge_capacity` in a higher rate message. It will only change slowly over multiple cycles and belongs in the INFO message IMO.

Certainly as you both point out `capacity_consumed` must be sent to capture the power monitor case. 
`Capacity_remaining` seems a bit redundant though ....

Reply:
Yes `capacity_remaining` is redundant. What was the original justification for `capacity_remaining` in mavlink? Is it so that someone can calculate full_charge_capacity without needing the SMART_BATTERY_INFO message?

Reply:
The battery status predates the SMART_BATTERY_INFO, which we have historically treated as "optional". I don't know the original reason we included this in the message since I think that GCSs mostly used voltage calculates to estimate capacity.

IF we are treating SMART_BATTERY_INFO as mandatory, albeit at lower rate, then it isn't necessary in MAVLink. I'll ask on discord, to confirm consensus with @tridge

Reply:
I'm not sure you could argue that `capacity_remaining` is redundant especially in a smart battery that can correct this value over time when in use. Reason why is that that `capacity_remaining` might even change while in use due to temperature changes, also a heavier load causes a battery heat up more quickly.
![https://batteryuniversity.com/article/bu-502-discharging-at-high-and-low-temperatures](https://batteryuniversity.com/img/content/discharge-voltage-temperature.jpg)

Furthermore for the end-user `capacity_remaining` is an easy to use indicator to know how much energy at this time could be retrieved even though it gets changed when in use.

Overall I think this message set looks fine to me and is a huge improvement to the older messages. And I hope we can this is in the DSDL repo sooner then later because I think not having a good battery message set in DroneCAN is hurting adoption of DroneCAN & Smart BMS at the current state.

Reply:
@PetervdPerk-NXP  I'm sorry for the delayed response.

Certainly, the `capacity_remaining` would change over time due to heavy loads or  temperature. 

But can the following arithmetic still apply?
`capacity_remaining = full_charge_capacity - capacity_consumed`

IE should or can the smart_battery's algorithms be updating the `full_charge_capacity` estimates.  And then place the burden on the end user (usually an autopilot) of the droneCAN message to compute `capacity_remaining`? 

Assuming we are sending `full_charge_capacity`  in the periodic message set every second to 5 seconds.

In mavlink, I agree having `capacity_remaining` in the message is important because it is easier for a human to understand. And no matter what we need to send both 2 out of the 3 items `full_charge_capacity`, `capacity_consumed`, or `capacity_remaning` since we are on a lossy link.  

Ideally, we send fewer bytes in the messages for DroneCAN? 

Of course `capacity_remaining` is only 4 bytes *10 Hz (320bits/sec per battery) . so I'm not opposed to keeping it in the high rate message if @dakejahl   is happy to keep it there too? And the droneCAN message would closely mirror mavlink that way. 

But if there are good ways to save bytes lets do it?

Reply:
I'm okay using the arithmetic you're proposing
`capacity_remaining = full_charge_capacity - capacity_consumed`

But we should change the description then that `full_charge_capacity` reflects this.

Regarding size of optimization the `20010.BatteryContinuous.uavcan` message, without capacity_remaining is 22 bytes, each DroneCAN CAN Frame can transmit up to 7 bytes, which means that you need 4 frames to transmit the data.
If we could optimize `20010.BatteryContinuous.uavcan` to be 21 bytes it would fit in 3 frames.

Reply:
Ahh yeah I forgot about the 7 byte payload requirement for DroneCAN. Thanks!

I was reading up here on multiframe transfers here, https://dronecan.github.io/Specification/4._CAN_bus_transport_layer/

It seems the first frame of multi-frame message has only 5 bytes for payload?
![image](https://user-images.githubusercontent.com/69225461/208765106-ba46967a-bbc2-4912-bcba-6b964d222c80.png)


So with 22 bytes of payload we must use 4 frames total?

Reply:
> I'm okay using the arithmetic you're proposing

   ```
   capacity_remaining = full_charge_capacity - capacity_consumed
   ```

Note that mavlink requires this to be true if [MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL](https://mavlink.io/en/messages/development.html#MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL) is set - i.e. in a smart battery the capacity consumed is always relative to full.

With a power monitor that flag is not set. In this case we are indicating capacity_consumed is "since boot". We _assume_ that the battery was full to start with, so you can use the calculation as shown. The difference though is that we know it might be wrong, and indeed will be on battery swaps etc.